test/extensions/quic_listeners/quiche/envoy_quic_dispatcher_test.cc

```diff
@@ -0,0 +1,226 @@
+#include "test/test_common/environment.h"
+#include "test/mocks/network/mocks.h"
+#include "test/test_common/utility.h"
+#include "test/test_common/network_utility.h"
+#include "extensions/quic_listeners/quiche/platform/envoy_quic_clock.h"
+#include "extensions/quic_listeners/quiche/envoy_quic_utils.h"
+#include "extensions/quic_listeners/quiche/envoy_quic_dispatcher.h"
+#include "extensions/quic_listeners/quiche/envoy_quic_fake_proof_source.h"
+#include "extensions/quic_listeners/quiche/envoy_quic_alarm_factory.h"
+#include "server/configuration_impl.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::Invoke;
+using testing::Return;
+using testing::ReturnRef;
+
+namespace quic {
+namespace test {
+class QuicDispatcherPeer {
+public:
+  static quic::QuicTimeWaitListManager* time_wait_list_manager(QuicDispatcher* dispatcher) {
+    return dispatcher->time_wait_list_manager_.get();
+  }
+};
+
+} // namespace test
+} // namespace quic
+
+namespace Envoy {
+namespace Quic {
+
+class EnvoyQuicDispatcherTest : public testing::TestWithParam<Network::Address::IpVersion>,
+                                protected Logger::Loggable<Logger::Id::main> {
+public:
+  EnvoyQuicDispatcherTest()
+      : version_(GetParam()), api_(Api::createApiForTest(time_system_)),
+        dispatcher_(api_->allocateDispatcher()), connection_helper_(*dispatcher_),
+        crypto_config_(quic::QuicCryptoServerConfig::TESTING, quic::QuicRandom::GetInstance(),
+                       std::make_unique<EnvoyQuicFakeProofSource>(),
+                       quic::KeyExchangeSource::Default()),
+        version_manager_(quic::CurrentSupportedVersions()),
+        listener_stats_({ALL_LISTENER_STATS(POOL_COUNTER(listener_config_.listenerScope()),
+                                            POOL_GAUGE(listener_config_.listenerScope()),
+                                            POOL_HISTOGRAM(listener_config_.listenerScope()))}),
+        connection_handler_(ENVOY_LOGGER(), *dispatcher_),
+        envoy_quic_dispatcher_(
+            &crypto_config_, &version_manager_,
+            std::make_unique<EnvoyQuicConnectionHelper>(*dispatcher_),
+            std::make_unique<EnvoyQuicAlarmFactory>(*dispatcher_, *connection_helper_.GetClock()),
+            quic::kQuicDefaultConnectionIdLength, connection_handler_, listener_config_,
+            listener_stats_) {
+    auto writer = new testing::NiceMock<quic::test::MockPacketWriter>();
+    envoy_quic_dispatcher_.InitializeWithWriter(writer);
+    EXPECT_CALL(*writer, WritePacket(_, _, _, _, _))
+        .WillRepeatedly(Return(quic::WriteResult(quic::WRITE_STATUS_OK, 0)));
+  }
+
+  void SetUp() override {
+    listen_socket_ = std::make_unique<Network::NetworkListenSocket<
+        Network::NetworkSocketTrait<Network::Address::SocketType::Datagram>>>(
+        Network::Test::getCanonicalLoopbackAddress(version_), nullptr, /*bind*/ true);
+    time_system_.sleep(std::chrono::milliseconds(100));
```

Comment:
Why do we need this sleep?

Reply:
SimulatedTimeSystemHelper initialize the systemTime to be 0. Advance the time can avoid QuicTime to be regarded as uninitialized.